File: sdks/typescript/src/__tests__/telemetry.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest';
import {
  dcpTelemetry,
  type TelemetryEvent,
  type DcpMetrics,
} from '../observability/telemetry.js';

describe('DcpTelemetry', () => {
  beforeEach(() => {
    dcpTelemetry.reset();
    dcpTelemetry.init({ enabled: false, exporterType: 'none' });
  });

  describe('disabled by default', () => {
    it('should not be enabled initially', () => {
      dcpTelemetry.reset();
      dcpTelemetry.init({ enabled: false });
      expect(dcpTelemetry.isEnabled).toBe(false);
    });

    it('should no-op all recording methods when disabled', () => {
      dcpTelemetry.init({ enabled: false });

      dcpTelemetry.recordSignLatency(10, 'ed25519');
      dcpTelemetry.recordVerifyLatency(5, 'ed25519');
      dcpTelemetry.recordKemLatency(20, 'encapsulate');
      dcpTelemetry.recordCheckpointLatency(15, 'routine');
      dcpTelemetry.recordBundleVerify(30, true, 'standard');
      dcpTelemetry.recordCacheHit();
      dcpTelemetry.recordCacheMiss();
      dcpTelemetry.recordA2ASession();
      dcpTelemetry.recordA2AMessage();
      dcpTelemetry.recordError('sign', 'fail');

      const metrics = dcpTelemetry.getMetrics();
      expect(metrics.signLatencyMs).toHaveLength(0);
      expect(metrics.verifyLatencyMs).toHaveLength(0);
      expect(metrics.kemLatencyMs).toHaveLength(0);
      expect(metrics.checkpointLatencyMs).toHaveLength(0);
      expect(metrics.bundleVerifyLatencyMs).toHaveLength(0);
      expect(metrics.cacheHits).toBe(0);
      expect(metrics.cacheMisses).toBe(0);
      expect(metrics.signaturesCreated).toBe(0);
      expect(metrics.signaturesVerified).toBe(0);
      expect(metrics.bundlesVerified).toBe(0);
      expect(metrics.a2aSessions).toBe(0);
      expect(metrics.a2aMessages).toBe(0);
      expect(metrics.errorsTotal).toBe(0);
    });

    it('should return empty span id when disabled', () => {
      dcpTelemetry.init({ enabled: false });
      const spanId = dcpTelemetry.startSpan('test-op');
      expect(spanId).toBe('');
    });

    it('should return 0 duration when ending span while disabled', () => {
      dcpTelemetry.init({ enabled: false });
      const duration = dcpTelemetry.endSpan('nonexistent');
      expect(duration).toBe(0);
    });
  });

  describe('init', () => {
    it('should enable telemetry with config', () => {
      dcpTelemetry.init({ serviceName: 'test-agent', enabled: true });
      expect(dcpTelemetry.isEnabled).toBe(true);
    });

    it('should emit init event when enabled', () => {
      const events: TelemetryEvent[] = [];
      dcpTelemetry.init({ enabled: true, serviceName: 'init-test' });
      dcpTelemetry.onEvent((e) => events.push(e));

      dcpTelemetry.reset();
      dcpTelemetry.init({ enabled: true, serviceName: 'init-test-2' });

      expect(events.some(e => e.type === 'init' && e.serviceName === 'init-test-2')).toBe(true);
    });

    it('should merge partial config with defaults', () => {
      dcpTelemetry.init({ enabled: true });
      expect(dcpTelemetry.isEnabled).toBe(true);
    });
  });

  describe('span lifecycle', () => {
    beforeEach(() => {
      dcpTelemetry.init({ enabled: true, serviceName: 'span-test' });
    });

    it('should create a span and return a non-empty id', () => {
      const spanId = dcpTelemetry.startSpan('crypto.sign', { algorithm: 'ed25519' });
      expect(spanId).toBeTruthy();
      expect(spanId.startsWith('crypto.sign-')).toBe(true);
    });

    it('should end a span and return positive duration', () => {
      const spanId = dcpTelemetry.startSpan('crypto.verify');
      const duration = dcpTelemetry.endSpan(spanId);
      expect(duration).toBeGreaterThanOrEqual(0);
    });

    it('should set error status on span', () => {
      const events: TelemetryEvent[] = [];
      dcpTelemetry.onEvent((e) => events.push(e));

      const spanId = dcpTelemetry.startSpan('failing-op');
      dcpTelemetry.endSpan(spanId, 'error', 'signature mismatch');

      const spanEvent = events.find(e => e.type === 'span');
      expect(spanEvent).toBeDefined();
      const span = (spanEvent as TelemetryEvent & { span: { status: string; error: string } }).span;
      expect(span.status).toBe('error');
      expect(span.error).toBe('signature mismatch');
    });

    it('should return 0 for unknown span id', () => {
      const duration = dcpTelemetry.endSpan('nonexistent-span-id');
      expect(duration).toBe(0);
    });

    it('should include service name in span attributes', () => {
      const events: TelemetryEvent[] = [];
      dcpTelemetry.onEvent((e) => events.push(e));

      const spanId = dcpTelemetry.startSpan('test-op');
      dcpTelemetry.endSpan(spanId);

      const spanEvent = events.find(e => e.type === 'span');
      const span = (spanEvent as TelemetryEvent & { span: { attributes: Record<string, unknown> } }).span;
      expect(span.attributes['dcp.service']).toBe('span-test');
    });
  });

  describe('metric recording', () => {
    beforeEach(() => {
      dcpTelemetry.init({ enabled: true, serviceName: 'metrics-test' });
    });

    it('should record sign latency', () => {
      dcpTelemetry.recordSignLatency(1.5, 'ed25519');
      dcpTelemetry.recordSignLatency(2.3, 'ml-dsa-65');
      const metrics = dcpTelemetry.getMetrics();
      expect(metrics.signLatencyMs).toEqual([1.5, 2.3]);
      expect(metrics.signaturesCreated).toBe(2);
    });

    it('should record verify latency', () => {
      dcpTelemetry.recordVerifyLatency(0.8, 'ed25519');
      const metrics = dcpTelemetry.getMetrics();
      expect(metrics.verifyLatencyMs).toEqual([0.8]);
      expect(metrics.signaturesVerified).toBe(1);
    });

    it('should record KEM latency', () => {
      dcpTelemetry.recordKemLatency(5.0, 'encapsulate');
      dcpTelemetry.recordKemLatency(4.2, 'decapsulate');
      const metrics = dcpTelemetry.getMetrics();
      expect(metrics.kemLatencyMs).toEqual([5.0, 4.2]);
    });

    it('should record checkpoint latency', () => {
      dcpTelemetry.recordCheckpointLatency(12.0, 'elevated');
      const metrics = dcpTelemetry.getMetrics();
      expect(metrics.checkpointLatencyMs).toEqual([12.0]);
    });

    it('should record bundle verify with tier distribution', () => {
      dcpTelemetry.recordBundleVerify(8.0, true, 'routine');
      dcpTelemetry.recordBundleVerify(15.0, true, 'standard');
      dcpTelemetry.recordBundleVerify(25.0, false, 'elevated');

      const metrics = dcpTelemetry.getMetrics();
      expect(metrics.bundleVerifyLatencyMs).toEqual([8.0, 15.0, 25.0]);
      expect(metrics.bundlesVerified).toBe(3);
      expect(metrics.tierDistribution['routine']).toBe(1);
      expect(metrics.tierDistribution['standard']).toBe(1);
      expect(metrics.tierDistribution['elevated']).toBe(1);
      expect(metrics.errorsTotal).toBe(1);
    });
  });

  describe('cache tracking', () => {
    beforeEach(() => {
      dcpTelemetry.init({ enabled: true, serviceName: 'cache-test' });
    });

    it('should track cache hits', () => {
      dcpTelemetry.recordCacheHit();
      dcpTelemetry.recordCacheHit();
      dcpTelemetry.recordCacheHit();
      expect(dcpTelemetry.getMetrics().cacheHits).toBe(3);
    });

    it('should track cache misses', () => {
      dcpTelemetry.recordCacheMiss();
      expect(dcpTelemetry.getMetrics().cacheMisses).toBe(1);
    });

    it('should compute cache hit rate in summary', () => {
      dcpTelemetry.recordCacheHit();
      dcpTelemetry.recordCacheHit();
      dcpTelemetry.recordCacheHit();
      dcpTelemetry.recordCacheMiss();

      const summary = dcpTelemetry.getMetricsSummary();
      expect(summary.cacheHitRate).toBe(0.75);
    });

    it('should return 0 cache hit rate when no cache operations', () => {
      const summary = dcpTelemetry.getMetricsSummary();
      expect(summary.cacheHitRate).toBe(0);
    });
  });

  describe('metrics summary with percentiles', () => {
    beforeEach(() => {
      dcpTelemetry.init({ enabled: true, serviceName: 'summary-test' });
    });

    it('should compute percentiles for recorded latencies', () => {
      for (let i = 1; i <= 100; i++) {
        dcpTelemetry.recordSignLatency(i, 'ed25519');
      }

      const summary = dcpTelemetry.getMetricsSummary();
      expect(summary.sign.count).toBe(100);
      expect(summary.sign.min).toBe(1);
      expect(summary.sign.max).toBe(100);
      expect(summary.sign.p50).toBe(51);
      expect(summary.sign.p95).toBe(96);
      expect(summary.sign.p99).toBe(100);
      expect(summary.sign.mean).toBeCloseTo(50.5);
    });

    it('should return zeroed stats for empty latency arrays', () => {
      const summary = dcpTelemetry.getMetricsSummary();
      expect(summary.sign).toEqual({
        count: 0, min: 0, max: 0, p50: 0, p95: 0, p99: 0, mean: 0,
      });
      expect(summary.verify).toEqual({
        count: 0, min: 0, max: 0, p50: 0, p95: 0, p99: 0, mean: 0,
      });
    });

    it('should include totals in summary', () => {
      dcpTelemetry.recordSignLatency(1, 'ed25519');
      dcpTelemetry.recordVerifyLatency(2, 'ed25519');
      dcpTelemetry.recordBundleVerify(3, true, 'routine');
      dcpTelemetry.recordError('sign', 'bad key');
      dcpTelemetry.recordA2ASession();
      dcpTelemetry.recordA2AMessage();

      const summary = dcpTelemetry.getMetricsSummary();
      expect(summary.totals.signaturesCreated).toBe(1);
      expect(summary.totals.signaturesVerified).toBe(1);
      expect(summary.totals.bundlesVerified).toBe(1);
      expect(summary.totals.errors).toBe(1);
      expect(summary.totals.a2aSessions).toBe(1);
      expect(summary.totals.a2aMessages).toBe(1);
    });

    it('should include tier distribution in summary', () => {
      dcpTelemetry.recordBundleVerify(5, true, 'routine');
      dcpTelemetry.recordBundleVerify(10, true, 'routine');
      dcpTelemetry.recordBundleVerify(15, true, 'maximum');

      const summary = dcpTelemetry.getMetricsSummary();
      expect(summary.tierDistribution['routine']).toBe(2);
      expect(summary.tierDistribution['maximum']).toBe(1);
    });
  });

  describe('event listeners', () => {
    beforeEach(() => {
      dcpTelemetry.init({ enabled: true, serviceName: 'listener-test' });
    });

    it('should register and receive events', () => {
      const events: TelemetryEvent[] = [];
      dcpTelemetry.onEvent((e) => events.push(e));

      dcpTelemetry.recordSignLatency(1.0, 'ed25519');

      expect(events).toHaveLength(1);
      expect(events[0].type).toBe('metric');
      expect(events[0].name).toBe('sign_latency_ms');
    });

    it('should allow unsubscribing from events', () => {
      const events: TelemetryEvent[] = [];
      const unsub = dcpTelemetry.onEvent((e) => events.push(e));

      dcpTelemetry.recordSignLatency(1.0, 'ed25519');
      expect(events).toHaveLength(1);

      unsub();
      dcpTelemetry.recordSignLatency(2.0, 'ed25519');
      expect(events).toHaveLength(1);
    });

    it('should not break if a listener throws', () => {
      dcpTelemetry.onEvent(() => { throw new Error('boom'); });
      const events: TelemetryEvent[] = [];
      dcpTelemetry.onEvent((e) => events.push(e));

      dcpTelemetry.recordSignLatency(1.0, 'ed25519');
      expect(events).toHaveLength(1);
    });

    it('should emit span events', () => {
      const events: TelemetryEvent[] = [];
      dcpTelemetry.onEvent((e) => events.push(e));

      const spanId = dcpTelemetry.startSpan('test');
      dcpTelemetry.endSpan(spanId);

      expect(events.some(e => e.type === 'span')).toBe(true);
    });
  });

  describe('error recording', () => {
    beforeEach(() => {
      dcpTelemetry.init({ enabled: true, serviceName: 'error-test' });
    });

    it('should increment error count', () => {
      dcpTelemetry.recordError('verify', 'invalid signature');
      dcpTelemetry.recordError('sign', 'key expired');

      expect(dcpTelemetry.getMetrics().errorsTotal).toBe(2);
    });

    it('should emit error events', () => {
      const events: TelemetryEvent[] = [];
      dcpTelemetry.onEvent((e) => events.push(e));

      dcpTelemetry.recordError('verify', 'bad hash');

      const errorEvent = events.find(e => e.type === 'error');
      expect(errorEvent).toBeDefined();
      expect(errorEvent!.operation).toBe('verify');
      expect(errorEvent!.error).toBe('bad hash');
    });
  });

  describe('reset', () => {
    it('should clear all metrics and spans', () => {
      dcpTelemetry.init({ enabled: true, serviceName: 'reset-test' });

      dcpTelemetry.recordSignLatency(1.0, 'ed25519');
      dcpTelemetry.recordVerifyLatency(2.0, 'ed25519');
      dcpTelemetry.recordCacheHit();
      dcpTelemetry.recordA2ASession();
      dcpTelemetry.startSpan('active-span');

      dcpTelemetry.reset();

      const metrics = dcpTelemetry.getMetrics();
      expect(metrics.signLatencyMs).toHaveLength(0);
      expect(metrics.verifyLatencyMs).toHaveLength(0);
      expect(metrics.cacheHits).toBe(0);
      expect(metrics.a2aSessions).toBe(0);
      expect(metrics.signaturesCreated).toBe(0);
      expect(metrics.signaturesVerified).toBe(0);
    });
  });

  describe('A2A session and message recording', () => {
    beforeEach(() => {
      dcpTelemetry.init({ enabled: true, serviceName: 'a2a-test' });
    });

    it('should track A2A sessions', () => {
      dcpTelemetry.recordA2ASession();
      dcpTelemetry.recordA2ASession();
      expect(dcpTelemetry.getMetrics().a2aSessions).toBe(2);
    });

    it('should track A2A messages', () => {
      dcpTelemetry.recordA2AMessage();
      dcpTelemetry.recordA2AMessage();
      dcpTelemetry.recordA2AMessage();
      expect(dcpTelemetry.getMetrics().a2aMessages).toBe(3);
    });

    it('should emit event for A2A sessions', () => {
      const events: TelemetryEvent[] = [];
      dcpTelemetry.onEvent((e) => events.push(e));

      dcpTelemetry.recordA2ASession();

      const sessionEvent = events.find(e => e.name === 'a2a_sessions_total');
      expect(sessionEvent).toBeDefined();
      expect(sessionEvent!.value).toBe(1);
    });
  });

  describe('console exporter', () => {
    it('should log to console when exporterType is console', () => {
      const consoleSpy = vi.spyOn(console, 'log').mockImplementation(() => {});

      dcpTelemetry.init({ enabled: true, serviceName: 'console-test', exporterType: 'console' });
      dcpTelemetry.recordSignLatency(1.0, 'ed25519');

      expect(consoleSpy).toHaveBeenCalled();
      const lastCall = consoleSpy.mock.calls.at(-1)?.[0] as string;
      expect(lastCall).toContain('[DCP-AI Telemetry]');
      expect(lastCall).toContain('sign_latency_ms');

      consoleSpy.mockRestore();
    });

    it('should not log to console when exporterType is none', () => {
      const consoleSpy = vi.spyOn(console, 'log').mockImplementation(() => {});

      dcpTelemetry.init({ enabled: true, serviceName: 'silent-test', exporterType: 'none' });
      dcpTelemetry.recordSignLatency(1.0, 'ed25519');

      const dcpCalls = consoleSpy.mock.calls.filter(
        c => typeof c[0] === 'string' && c[0].includes('[DCP-AI Telemetry]'),
      );
      expect(dcpCalls).toHaveLength(0);

      consoleSpy.mockRestore();
    });
  });

  describe('OTLP exporter', () => {
    it('surfaces a helpful init error when OTel packages are missing', async () => {
      // The test env does not have @opentelemetry/* installed as real deps, so
      // initOtlpBridge() should fail, emit a telemetry error event, and leave
      // the exporter in a safe no-op state.
      dcpTelemetry.reset();
      const events: TelemetryEvent[] = [];
      const unsubscribe = dcpTelemetry.onEvent(e => events.push(e));

      const result = dcpTelemetry.init({
        enabled: true,
        serviceName: 'otlp-missing-deps',
        exporterType: 'otlp',
        otlpEndpoint: 'http://localhost:4318',
      });
      // init() returns a promise when exporterType === 'otlp'
      expect(result).toBeInstanceOf(Promise);
      await result;

      const errorEvents = events.filter(e => e.type === 'error');
      expect(errorEvents.length).toBeGreaterThan(0);
      const firstError = errorEvents[0] as unknown as { error?: string };
      expect(String(firstError.error ?? '')).toMatch(/opentelemetry/i);

      // And the app keeps working — listener-based telemetry still fires.
      dcpTelemetry.recordSignLatency(3.5, 'ed25519');
      const metricEvents = events.filter(e => e.type === 'metric');
      expect(metricEvents.length).toBeGreaterThan(0);

      unsubscribe();
    });

    it('does not attempt OTLP init when exporterType is console', () => {
      dcpTelemetry.reset();
      const result = dcpTelemetry.init({
        enabled: true,
        serviceName: 'console-only',
        exporterType: 'console',
      });
      // Synchronous for non-OTLP
      expect(result).toBeUndefined();
    });
  });
});

File: sdks/go/dcp/observability/telemetry_test.go
package observability_test

import (
	"sync"
	"testing"

	"github.com/dcp-ai-protocol/dcp-ai/sdks/go/v2/dcp/observability"
)

func fresh(t *testing.T) *observability.Telemetry {
	t.Helper()
	return observability.New()
}

func TestDisabledByDefault(t *testing.T) {
	tel := fresh(t)
	if tel.IsEnabled() {
		t.Fatal("telemetry should be disabled by default")
	}
	tel.RecordSignLatency(10, "ed25519")
	tel.RecordVerifyLatency(5, "ed25519")
	tel.RecordBundleVerify(30, true, "standard")
	s := tel.GetMetricsSummary()
	if s.Sign.Count != 0 {
		t.Fatalf("expected sign count 0, got %d", s.Sign.Count)
	}
	if s.Totals.SignaturesCreated != 0 {
		t.Fatalf("expected 0 signatures created, got %d", s.Totals.SignaturesCreated)
	}
}

func TestInitEmitsInitEvent(t *testing.T) {
	tel := fresh(t)
	var mu sync.Mutex
	var inits []string
	unsub := tel.OnEvent(func(e observability.Event) {
		if e.Type == observability.EventInit {
			mu.Lock()
			inits = append(inits, e.ServiceName)
			mu.Unlock()
		}
	})
	defer unsub()
	tel.Init(observability.Config{ServiceName: "svc", Enabled: true, ExporterType: observability.ExporterNone})
	mu.Lock()
	defer mu.Unlock()
	if len(inits) != 1 || inits[0] != "svc" {
		t.Fatalf("expected init event with svc, got %#v", inits)
	}
}

func TestSignAggregatesAndPercentiles(t *testing.T) {
	tel := fresh(t)
	tel.Init(observability.Config{Enabled: true})
	for v := 1; v <= 10; v++ {
		tel.RecordSignLatency(float64(v), "ed25519")
	}
	s := tel.GetMetricsSummary()
	if s.Sign.Count != 10 {
		t.Fatalf("expected 10, got %d", s.Sign.Count)
	}
	if s.Sign.Min != 1 || s.Sign.Max != 10 || s.Sign.Mean != 5.5 {
		t.Fatalf("unexpected stats: %#v", s.Sign)
	}
	if s.Sign.P50 != 6 {
		t.Fatalf("expected p50 = 6, got %v", s.Sign.P50)
	}
	if s.Totals.SignaturesCreated != 10 {
		t.Fatalf("expected 10 sigs created, got %d", s.Totals.SignaturesCreated)
	}
}

func TestTierDistributionGrows(t *testing.T) {
	tel := fresh(t)
	tel.Init(observability.Config{Enabled: true})
	tel.RecordBundleVerify(1, true, "routine")
	tel.RecordBundleVerify(1, true, "routine")
	tel.RecordBundleVerify(1, true, "elevated")
	s := tel.GetMetricsSummary()
	if s.TierDistribution["routine"] != 2 {
		t.Fatalf("expected routine = 2, got %d", s.TierDistribution["routine"])
	}
	if s.TierDistribution["elevated"] != 1 {
		t.Fatalf("expected elevated = 1, got %d", s.TierDistribution["elevated"])
	}
	if s.TierDistribution["maximum"] != 0 {
		t.Fatalf("expected maximum = 0, got %d", s.TierDistribution["maximum"])
	}
}

func TestSpanLifecycle(t *testing.T) {
	tel := fresh(t)
	tel.Init(observability.Config{Enabled: true})
	var mu sync.Mutex
	var spans []string
	unsub := tel.OnEvent(func(e observability.Event) {
		if e.Type == observability.EventSpan && e.Span != nil {
			mu.Lock()
			spans = append(spans, e.Span.Name)
			mu.Unlock()
		}
	})
	defer unsub()
	id := tel.StartSpan("dcp.test", map[string]interface{}{"n": 1})
	if id == "" {
		t.Fatal("expected span id")
	}
	d := tel.EndSpan(id)
	if d < 0 {
		t.Fatalf("duration should be non-negative, got %v", d)
	}
	mu.Lock()
	defer mu.Unlock()
	if len(spans) != 1 || spans[0] != "dcp.test" {
		t.Fatalf("unexpected span events: %#v", spans)
	}
}

func TestEndUnknownSpanIsNoop(t *testing.T) {
	tel := fresh(t)
	tel.Init(observability.Config{Enabled: true})
	if tel.EndSpan("nope") != 0 {
		t.Fatal("expected 0 for unknown span")
	}
}

func TestListenerPanicDoesNotBreakEmit(t *testing.T) {
	tel := fresh(t)
	tel.Init(observability.Config{Enabled: true})
	unsub := tel.OnEvent(func(_ observability.Event) {
		panic("listener bug")
	})
	defer unsub()
	// Must not propagate.
	tel.RecordSignLatency(1, "ed25519")
}

func TestCacheHitRate(t *testing.T) {
	tel := fresh(t)
	tel.Init(observability.Config{Enabled: true})
	tel.RecordCacheHit()
	tel.RecordCacheHit()
	tel.RecordCacheHit()
	tel.RecordCacheMiss()
	s := tel.GetMetricsSummary()
	if s.CacheHitRate != 0.75 {
		t.Fatalf("expected 0.75, got %v", s.CacheHitRate)
	}
}

func TestErrorIncrementsTotals(t *testing.T) {
	tel := fresh(t)
	tel.Init(observability.Config{Enabled: true})
	var got string
	unsub := tel.OnEvent(func(e observability.Event) {
		if e.Type == observability.EventError {
			got = e.Operation
		}
	})
	defer unsub()
	tel.RecordError("sign", "boom")
	if got != "sign" {
		t.Fatalf("expected sign, got %q", got)
	}
	if tel.GetMetricsSummary().Totals.Errors != 1 {
		t.Fatalf("expected 1 error, got %d", tel.GetMetricsSummary().Totals.Errors)
	}
}

func TestOTLPRequestedWithoutBuildTagSurfacesError(t *testing.T) {
	tel := fresh(t)
	var got string
	unsub := tel.OnEvent(func(e observability.Event) {
		if e.Type == observability.EventError && e.Operation == "otlp_init" {
			got = e.Error
		}
	})
	defer unsub()
	tel.Init(observability.Config{
		ServiceName:  "otlp-missing",
		Enabled:      true,
		ExporterType: observability.ExporterOTLP,
		OTLPEndpoint: "http://localhost:4318",
	})
	if got == "" {
		t.Fatal("expected otlp_init error event when built without -tags otlp")
	}
}
